autogluon/utils/tabular/ml/models/lr/lr_model.py

```diff
@@ -0,0 +1,206 @@
+                 feature_types_metadata=None, debug=0):
+        super().__init__(path=path, name=name, problem_type=problem_type, objective_func=objective_func, hyperparameters=hyperparameters, features=features,
+                         feature_types_metadata=feature_types_metadata, debug=debug)
+        self.types_of_features = None
+        self.pipeline = None
+        self.cat_one_hot = None
+
+    def tokenize(self, s):
+        return re.split('[ ]+', s)
+
+    def _get_types_of_features(self, df):
+        """ Returns dict with keys: : 'continuous', 'skewed', 'onehot', 'embed', 'language', values = ordered list of feature-names falling into each category.
+            Each value is a list of feature-names corresponding to columns in original dataframe.
+            TODO: ensure features with zero variance have already been removed before this function is called.
+        """
+        if self.types_of_features is not None:
+            Warning("Attempting to _get_types_of_features for TabularNeuralNetModel, but previously already did this.")
+        categorical_featnames = self.__get_feature_type_if_present('object') + self.__get_feature_type_if_present('bool')
+        continuous_featnames = self.__get_feature_type_if_present('float') + self.__get_feature_type_if_present('int') + self.__get_feature_type_if_present(
+            'datetime')
+        language_featnames = self.feature_types_metadata['nlp']
+        valid_features = categorical_featnames + continuous_featnames + language_featnames
+        if len(categorical_featnames) + len(continuous_featnames) \
+                + len(language_featnames) \
+                != df.shape[1]:
+            unknown_features = [feature for feature in df.columns if feature not in valid_features]
+            df = df.drop(columns=unknown_features)
+            self.features = list(df.columns)
+
+        types_of_features = {'continuous': [], 'skewed': [], 'onehot': [], 'language': []}
+        return self._select_features(df, types_of_features, categorical_featnames, language_featnames, continuous_featnames)
+
+    def _select_features(self, df, types_of_features, categorical_featnames, language_featnames, continuous_featnames):
+        raise NotImplementedError()
+
+    def __get_feature_type_if_present(self, feature_type):
+        """ Returns crude categorization of feature types """
+        return self.feature_types_metadata[feature_type] if feature_type in self.feature_types_metadata else []
+
+    # TODO: handle collinear features - they will impact results quality
+    def preprocess(self, X: DataFrame, is_train=False, vect_max_features=1000):
+        X = X.copy()
+        feature_types = self._get_types_of_features(X)
+        logger.log(15, "Applying model-specific pre-processing")
+        logger.log(15, " - input shape %s" % str(X.shape))
+        if is_train:
```

Comment:
Can you move this to a preprocess_train() function? Longer term I want to consolidate this logic for stateful preprocessors and having it separated will simplify things.

Reply:
Done